module/zfs/dsl_synctask.c

```diff
@@ -86,10 +86,15 @@ dsl_sync_task_common(const char *pool, dsl_checkfunc_t *checkfunc,
 
 	dmu_tx_commit(tx);
 
-	if (sigfunc != NULL && txg_wait_synced_sig(dp, dst.dst_txg)) {
-		/* current contract is to call func once */
-		sigfunc(arg, tx);
-		sigfunc = NULL;	/* in case we're performing an EAGAIN retry */
+	if (sigfunc != NULL) {
+		err = txg_wait_synced_flags(dp, dst.dst_txg, TXG_WAIT_SIGNAL);
+		if (err != 0) {
+			VERIFY3U(err, ==, EINTR);
+			/* current contract is to call func once */
+			sigfunc(arg, tx);
+			/* in case we're performing an EAGAIN retry */
+			sigfunc = NULL;
+		}
 	}
 	txg_wait_synced(dp, dst.dst_txg);
```

Comment:
Its been like that before, but second `txg_wait_synced()` call seems unneeded if `err == 0`.

Reply:
Ahh yeah, but only for the sigfunc case; if there's no sigfunc, then the error will be zero. Bit awkward, but fixed I think.